### ISSUE TITLE
Transactional support for Kafka consume-transform-produce workflows

### ISSUE THREAD

**Issue by seglo** [akka/reactive-kafka#369]
Kafka announced several new features in its 0.11.0.0 release that collectively allow for exactly-once semantics, or "effective-once" depending on your opinions on the subject.  Semantics aside, there are several new Kafka client features that would be nice to expose in reactive-kafka's API.

Exactly-once delivery on the Kafka producer was impossible to guarantee before the release of 0.11.0.0 because there was a possibility that a producer could send a duplicate message in a retry and fail before it received an acknowledgement.

Kafka achieves exactly-once by making a number of changes to Kafka broker and client code.

* Idempotent producers. When `ENABLE_IDEMPOTENCE` is enabled in the producer config it will coordinate with the broker to ensure that retried messages aren't duplicated.
* Transactions have been introduced to guarantee exactly once message delivery in a `consume->transform->produce` workflow. This is a common operation involving consuming from a Kafka topic, processing, and producing to a new topic. To make transactions a reality in Kafka 0.11 the following changes were made:
  * Idempotent producers
  * Transaction Coordinator: New logical component on broker 
  * Message schema changes: New concepts such as Transaction ID, Producer ID, and transaction sequence numbers.

Supporting these features will help close the gap between Kafka streams and Akka streams w/ reactive-kafka for stateless streaming/transformation workflows.  I would like to work on a PR to implement these new features of the Kafka client.

Kafka Improvement Proposals (KIP) for Transaction Support
* [KIP-98: Exactly Once Delivery and Transactional Messaging](https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging)
* [KIP-129: Streams Exactly-Once Semantics](https://cwiki.apache.org/confluence/display/KAFKA/KIP-129%3A+Streams+Exactly-Once+Semantics)

**Comment by ktoso** [comment]
I wonder how much it can be incorporated -- looking forward to a PR then @seglo !

**Comment by seglo** [comment]
Great, thanks! I've used these features with the underlying Kafka client libs.  I'm optimistic it can be incorporated.  I'll start working on it this week.

**Comment by mjkrumlauf** [comment]
@seglo how is this progressing?

**Comment by seglo** [comment]
@mjkrumlauf It hasn't progressed very far unfortunately.  I should have the bandwidth to work on this late next week.  If that changes I'll make a reminder to update this ticket so I don't hold back anyone else who may want to take a crack at it.

**Comment by mjkrumlauf** [comment]
@seglo Thanks for the quick response.  I understand the part about not having bandwidth!

**Comment by rafalmag** [comment]
I analyzed this a little bit.
The most important use case was ability to:
- consume message from input topic
- transform it to output message(s)
- in a single transaction commit the input message offset and output message(s)

The big issue is the "zombie fencing" - how to assign the "transaction id" and handle rebalancing.
Kafka broker can detect when consumer is not available and triggers the consumers rebalancing, but the message could be still processed and the producer could still try to commit the message offset and produce the output messages.
It should be prevented by "zombie fencing" - transaction ids + epochs. 

https://www.confluent.io/blog/transactions-apache-kafka/
"Exactly Once Delivery and Transactional Messaging in Kafka":
https://docs.google.com/document/d/11Jqy_GjUGtdXJK94XGsEIK7CP1SnQGdp2eF0wSw9ra8/edit?usp=sharing


In akka streams world this can be achieved using:
My rough idea was to use `Consumer.plainPartitionedSource` to get new stream after each re-balancing,
for each of the subsources attach custom "flow" that would produce the output messages
attach custom producer that would consume the output messages and open the transaction, send the output messages and commit the offset of the input messages.

The hard part that "initTransaction" should happen ideally in the same thread as kind of a callback to the repartitioning (maybe as a callback from `KafkaConsumerActor`).
The second part is that when the repartitioning happens and the partition is reassign to different consumer - the producer should be synchronously cancelled (also ideally as an action from `KafkaConsumerActor`).

From architecture point of view right now I cannot see how Source and Sink could be decoupled from each other, so from API point of view it might looks like this:

```
def withTransactions(consumerSettings: ConsumerSettings, producerSettings: ProducerSettings, flow: Flow[ConsumerRecord, Seq[ProducerRecord], Mat]): RunnableGraph[Mat]
```

**Comment by mjkrumlauf** [comment]
I am not familiar with the rebalancing difficulties you mention but I think if the Source and Sink are not able to be decoupled this would extremely limit the usefulness of transactional producers and consumers in our context.

**Comment by seglo** [comment]
I've been studying your analysis @rafalmag and I have a few questions.  I would like to clarify how an app using transactions would run.  

I'm making the assumption that for a given topic partition subscription there can only be _one_ active consume->transform->produce application instance.  Unlike consumer group balancing which could distribute a subscription across many consumer application instances in the same consumer group (where # application instances <= total number of partitions), we can't do the same with a transactional consume->transform-> produce workflow because only one application instance per `transaction.id` is allowed.  Any other application instances with the same `transaction.id` (zombie instances), would be fenced off.  This is a fundamental rule of transactions in Kafka.  Do you agree with this explanation?

Assuming the above is true, then the only time we would want to rebalance is if the first application instance is offline.  This is effectively a master/slave deployment architecture.

In your design proposal you include this point:

> create and attach custom producer that would consume the output messages and open the transaction, send the output messages and commit the offset of the input messages. The producer could set the trancasctionId that would include the partition ids assigned to the consumer (#387 related).

Are you suggesting here that somehow reactive-kafka create a subset of the larger topicpartition subscription (basically what the consumer group coordinator does) and encode that information into the transaction.id so we can parallelize the transactional consume->transform->produce workflow?

**Comment by seglo** [comment]
Just an update that I've been working on some ideas. Should have something to share next week.

**Comment by seglo** [comment]
Hey.  I've run into a problem that maybe some of you more familiar with akka streams internals could assist with.  The `KafkaProducer` is thread safe and most standard Kafka operations don't care that it's shared.  However when using the producer with a transactional workflow we have to maintain a strict lifecycle to avoid entering an inconsistent state.  

The Kafka transactional workflow with reactive-kafka would be to call `KafkaProducer.initTransaction` at `Source` instantiation to assure a producer with the same transactional id is not running.  The `KafkaConsumerActor` begins pumping messages out to one of our `GraphStateLogic` implementations.  A new transaction begins using `KafkaProducer.beginTransaction`.  The user transforms consumed messages and we call `KafkaProducer.send`.  Assuming we're not batching records, then once the message is successfully produced we acknowledge Kafka transaction coordinator and consumer group coordinator to which records have been consumed and commit the transaction with `KafkaProducer.sendOffsetsToTransaction` and `KafkaProducer.commitTransaction`.  After this is a success we're ready to begin a new transaction.

This above workflow is simple enough, but it imposes a single threaded model.  For example, I cannot use the same producer to send a record while its in the process of committing a previous transaction.  If a transaction is in the process of being aborted, then you can't begin a new transaction or send records to a different transaction.  You can see that there would be a number of scenarios where this model breaks when transaction operations are called asynchronously.

If the stream is entirely fused and contains no async operations then it may be possible to model this workflow by forcing it to effectively be a single threaded loop, but I don't see how the library would provide much value in enforcing these constraints.

The way KafkaStreams implements transactions is by assigning a single producer to a thread and not sharing it.  Each thread is dedicated to processing 1 or more partitions from the consumer subscription.  Each thread manages the state of producer transaction carefully so it doesn't get into an inconsistent state.

Is there an easy way to model an akka stream in the same way?  If I had control over the threading (or actor) model I could manage the state of the producer and route the messages that use it to the right thread/actor based on some kind of partitioning model.  Is this something that could be done with an akka streams `Materializer`?

I have a working branch with my work so far that I could push, but I'll need to clean it up first.  I thought it was important to raise the issues I've been encountering first to see what others might suggest.

**Comment by patriknw** [comment]
Relying on that different stages are fused (running single threaded) is not something I'd recommend. Sounds like this specific feature should be implemented as a single GraphStage or Actor.

**Comment by seglo** [comment]
Yes.  My first implementation attempt relied on my own implementations using the existing `Committable` and `Committer` types with no changes to the existing `ConsumerStage` or `ProducerStage`, but this won't work.  Although there are similarities between transactional commits and the Kafka managed consumer group offset commit semantics there are enough differences to warrant at least a modified `ProducerStage`.  I need to maintain some state about running transactions in the `ProducerStage` to prevent trampling of the transactional workflow for its one producer instance.  Once I get that working I'll think about how it can be scaled out.

**Comment by rafalmag** [comment]
When I was reading and analyzing the debug logs of Kafka Streams - they are using single (same!) thread for a consumer-producer coordination too. So when consumer figures out that it has been kicked out from consumption of given topic-partition then the producer is not sending the output message (nor committing offset). In order to achieve the same in akka-streams world - I think the same actor instance should to be used and act as a consumer and a producer. Then as it would control the sink - in case of re-balancing and being kicked out of the topic-partition it could cancel the flow from the sink side.

In order to scale to multiple threads:
- then multiple actors should be used
- As there are no assumptions regarding the order of the messages from different partitions - then it seems natural to handle processing of messages from different partitions in "parallel" flows.
That leaded to an idea that the actors should be bound to input topic-partitions - something similar to
`Consumer.plainPartitionedSource` - but unfortunately the part about direct (and synchronous) notification to sink is missing in the API.

**Comment by seglo** [comment]
Agreed, partitioning the transactions by a subset of partitions is the best way to scale out, but I don't think this should be part of transactional support in reactive-kafka.  If we attempted to manage this functionality internally then in order to scale we would need to subdivide the configured consumer group subscription ourselves.  The producer requires a unique consumer group id when committing offsets back to the transaction coordinator and consumer group coordinator _before_ committing the transaction itself.  AFAIK we can't sub divide an existing consumer group to commit multiple transactions (1:many).  I think this can be left to the user to do by having them run multiple reactive-kafka apps in parallel, each with its own unique/non-overlapping consumer group.

The approach I'm working on now is to create a new producer stage that's initialized with a batching parameter.  The stage will then initialize transactions (once), begin a transaction, accept and send messages until a batching parameter is met.  When the commit takes place the stage cannot produce any more messages until it's been complete and a new transaction begins.  In the event of a failure we restart the stream from the last committed consumer offsets.  

I'm also toying with a way of handling this failure internally for specified user defined exceptions.  Kafka Streams doesn't currently support this and leaves it to the user, but they acknowledge in the EoS for Kafka Stream KIP that this could be a possibility in the future.  For now, I think if we include a documentation example of how to restart the stream outside reactive-kafka it would be good enough.

**Comment by seglo** [comment]
For anyone following this thread I've put a PR ^ https://github.com/akka/reactive-kafka/pull/420

**Comment by rafalmag** [comment]
Here is a little bit of the research that we did in our team:

## Processing guarantee
 
We use _exactly-once-processing_.
 
To achieve this:
- our consumers consume with `READ_COMMITTED` isolation level 
- our producers are idempotent and transactional
- we use Kafka's zombie fencing mechanism (**IN PROGRESS**)
 
Read up:                                                                                                                                                                                                                           
- [Exactly-once Semantics are Possible: Here’s How Kafka Does it](https://www.confluent.io/blog/exactly-once-semantics-are-possible-heres-how-apache-kafka-does-it/)
- [Chain Services with Exactly-Once Guarantees](https://www.confluent.io/blog/chain-services-exactly-guarantees/)
- [Transactions in Apache Kafka](https://www.confluent.io/blog/transactions-apache-kafka/)
 

### Zombie-fencing

Zombie process with _consume-process-publish_ model is a process which consumer "lost" a topic-partition for which
it is still processing a record. Loosing topic-partition assigned by Kafka can happen because of network issues or any
change in consumer group, like adding or removing group member, which then triggers group rebalancing, i.e. reassignment of topic-partitions.

With transactional exactly-once-processing approach, we commit consumer offsets, not using the consumer, but instead using
the same transactional producer that we use for publishing.
So if we process event A1 which maps to event B1, then we publish in the same transaction B1 and consumed offset for A1.
This is good, because we only "ACK" processing A1 if B1 is successfully published. But if we loose a partition, and it gets
reassigned to another member, before we finish (or even start) transaction committing offset for A1, then that other member can pick up
A1 and process it as well. This will result in duplicated record B1 and will break exactly-once-processing.
 
To avoid this, Kafka [implements zombie fencing mechanism](https://www.confluent.io/blog/transactions-apache-kafka/):
 
> We solve the problem of zombie instances by requiring that each transactional producer be assigned a unique identifier called the transactional.id. This is used to identify the same producer instance across process restarts.
>
> The API requires that the first operation of a transactional producer should be to explicitly register its transactional.id with the Kafka cluster. When it does so, the Kafka broker checks for open transactions with the given transactional.id and completes them. It also increments an epoch associated with the transactional.id. The epoch is an internal piece of metadata stored for every transactional.id.
>
> Once the epoch is bumped, any producers with same transactional.id and an older epoch are considered zombies and are fenced off, ie. future transactional writes from those producers are rejected.

Sounds easy, but picking the right `transactional.id` is actually a bit tricky:
 
> The key to fencing out zombies properly is to ensure that the input topics and partitions in the read-process-write cycle is always the same for a given transactional.id. If this isn’t true, then it is possible for some messages to leak through the fencing provided by transactions.
>
> Practically, one would either have to store the mapping between input partitions and transactional.ids in an external store, or have some static encoding of it.

Kafka Streams has this built-in using static encoding, and for our stateless steps with Akka Streams we decided to model it after Kafka Streams.
The simplest _static_ encoding for this is one-to-one mapping between input topic-partition and producer (`transactional-id`).
So if we have 2 instances of step `foo: a => b` consuming from topic `t-0` with 4 partitions, than we can have:
 
```
foo-0:
  t-0, p-0 ---> consumer[group.id = foo] ---> a0[t-0, p-0] --> b0[a0] --> producer[transactional.id = foo-t-0-0]
  t-0, p-1 -/                             \-> a1[t-0, p-1] --> b1[a1] --> producer[transactional.id = foo-t-0-1]         
 
foo-1:
  t-0, p-2 ---> consumer[group.id = foo] ---> a2[t-0, p-2] --> b2[a2] --> producer[transactional.id = foo-t-0-2]
  t-0, p-3 -/                             \-> a3[t-0, p-3] --> b3[a3] --> producer[transactional.id = foo-t-0-3]
```
 
- one consumer with the same `group.id` per step instance (process or thread)
- one producer per input topic-partition with statically encoded `transactional.id`
- set of producers in every instance reflects current topic-partition assignment, and changes after reassignment
- output topic and partition are irrelevant

**Comment by ennru** [comment]
Implemented by #420
